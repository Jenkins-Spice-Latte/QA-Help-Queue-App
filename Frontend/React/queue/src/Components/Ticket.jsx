
import { CustomInput, Collapse, Button, CardBody, Card, Modal, ModalHeader, ModalBody, ModalFooter,  Form, FormGroup, Label, InputGroup, InputGroupText, InputGroupAddon, Input } from 'reactstrap';
import { useState } from 'react';
import axios from 'axios';
import { BsChevronDown, BsChevronUp, BsClockFill } from "react-icons/bs";
import { FaRegCheckCircle, FaCheckCircle } from "react-icons/fa";

const Ticket = (props) => {

    const item = props.item;
    const className = props.className;


    let TopicCheck1 = false;
    let TopicCheck2 = false;
    let TopicCheck3 = false;
    let TopicCheck4 = false;
    let TopicCheck5 = false;
    let UrgencyCheck1 = false;
    let UrgencyCheck2 = false;
    let UrgencyCheck3 = false;
    let UrgencyCheck4 = false;
    let UrgencyCheck5 = false;
    const decidedComp = item.complete;
    const  decidedTopic = item.topic;
    const  decidedUrgency = item.urgency;
    let Urgencylevel;
    let checkAuth;
    let checkTitle;
    let checkDesc;
    let urgencyCheck;
    let topicCheck;
    
    const [modal, setModal] = useState(false);
    const [isOpen, setIsOpen] = useState(false);
    const [authorSt, setAuthor] = useState(item.author);
    const [completeSt, setComplete] = useState("");
    const [descriptionSt, setDescription] = useState(item.description);
    const [timeSt, setTime] = useState('');
    const [titleSt, setTitle] = useState(item.title);
    const [topicSt, setTopic] = useState(item.topic);
    const [urgencySt, setUrgency] = useState(item.urgency);

    
    let date = new Date(item.time_created);
    let hours = date.getHours();
    let mins = "" + date.getMinutes();

    let btn;
    let tickBtn;
    let priorityBtn;
    let disabled;
    let markasdoneDisabled;
    
    if(props.mode === "Trainer mode"){
      disabled = false;
    } else{
      disabled = true;
    }

    const isUpdateEnabled = authorSt.length > 0 && titleSt.length > 0 && descriptionSt.length > 0 && topicSt.length > 0 && urgencySt.length > 0;


    const isEnabled = item.complete;

    if(isEnabled){
      markasdoneDisabled = true;
    } else if(props.mode === "Trainee mode") {
      markasdoneDisabled = true;
    } else{
      markasdoneDisabled = false;
    }


    const expand = () => setIsOpen(!isOpen);

    

    const mark = () => {

      let ticket = {
        author: authorSt,
        complete: true,
        description: descriptionSt,
        time_created: timeSt,
        title: titleSt,
        topic: topicSt,
        urgency: urgencySt
      };


      axios.put("http://localhost:8903/update/"+item.ticketID, ticket)
      .then(response => {
        console.log(response.data);
        props.switchLoaded()
      });  
    }

    const deleteT = () => {
      axios.delete("http://localhost:8904/delete/"+item.ticketID)
      .then(response => {
        console.log(response.data);
        props.switchLoaded()
      });
    }

    const handleSubmit = event => {
      event.preventDefault();
  
      let ticket = {
        author: authorSt,
        complete: completeSt,
        description: descriptionSt,
        time_created: timeSt,
        title: titleSt,
        topic: topicSt,
        urgency: urgencySt
      };
      
      axios.put("http://localhost:8903/update/"+item.ticketID,  ticket)
        .then(res => {
          console.log(res);
          console.log(res.data);
          props.switchLoaded()
        })
    }

    const toggle = () => setModal(!modal);


      if (isOpen) {
        btn = <BsChevronUp className="contractQueueIc" onClick={expand}/>;
      } else {
        btn = <BsChevronDown className="expandQueueIc" onClick={expand}/>;
      }
  
      if (decidedComp) {
        tickBtn = <FaCheckCircle className="completedIc"/>;
      } else {
        tickBtn = <FaRegCheckCircle className="uncompletedIc"/>;
      }
  
      if(decidedUrgency === 1){
        priorityBtn = <BsClockFill className="mostUrgIcQ"/>
      } else if(decidedUrgency === 2){
        priorityBtn = <BsClockFill className="secMostUrgIcQ"/>
      }else if(decidedUrgency === 3){
        priorityBtn = <BsClockFill className="middleUrgIcQ"/>
      }else if(decidedUrgency === 4){
        priorityBtn = <BsClockFill className="secLeastUrgIcQ"/>
      }else if(decidedUrgency === 5){
        priorityBtn = <BsClockFill className="leastUrgIcQ"/>
      }

      if(decidedUrgency === 1){
        UrgencyCheck1 = true;
        Urgencylevel = "Most urgent";
      } else if(decidedUrgency === 2){
        UrgencyCheck2 = true;
        Urgencylevel = "Very urgent";
      }else if(decidedUrgency === 3){
        UrgencyCheck3 = true;
        Urgencylevel = "Slightly urgent";
      }else if(decidedUrgency === 4){
        UrgencyCheck4 = true;
        Urgencylevel = "Less urgent";
      }else if(decidedUrgency === 5){
        UrgencyCheck5 = true;
        Urgencylevel = "Least urgent";
      }

      if(decidedTopic === "Dev Ops"){
        TopicCheck1 = true;
      } else if(decidedTopic === "General"){
        TopicCheck2 = true;
      }else if(decidedTopic === "Back-end"){
        TopicCheck3 = true;
      }else if(decidedTopic === "Front-end"){
        TopicCheck4 = true;
      }else if(decidedTopic === "Software"){
        TopicCheck5 = true;
      }

      if(urgencySt !== ""){
        urgencyCheck = <p id="createUrgencyCheck">Urgency selected</p> 
      }
    
      if(topicSt !== ""){
        topicCheck = <p id="createTopicCheck">Topic selected</p> 
      }
      
      if(authorSt === ""){
        checkAuth = <Input type="text" name="author" id="author" value={authorSt} onChange={(e) => setAuthor(e.target.value)} placeholder="Author name"/>
        
      } else{
        checkAuth = <Input valid type="text" name="author" id="author" value={authorSt} onChange={(e) => setAuthor(e.target.value)} placeholder="Author name"/>
      }
    
      if(titleSt === ""){
        checkTitle = <Input type="text" name="title" id="title" value={titleSt} onChange={(e) => setTitle(e.target.value)} placeholder="Ticket title"/>
        
      } else{
        checkTitle = <Input valid type="text" name="title" id="title" value={titleSt} onChange={(e) => setTitle(e.target.value)} placeholder="Ticket title"/>
      }
    
      if(descriptionSt === ""){
        checkDesc = <Input type="textarea" name="description" value={descriptionSt} onChange={(e) => setDescription(e.target.value)} id="description" placeholder="Description" />
        
      } else{
        checkDesc = <Input valid type="textarea" name="description" value={descriptionSt} onChange={(e) => setDescription(e.target.value)} id="description" placeholder="Description" />
      }


    return (
        <div className="ticket_div" key={item.id}>
              <p className="ticket_comp title_comp">{item.title} ({item.topic}) </p>    
              {btn}
              {tickBtn}
              {priorityBtn}
              
              <Collapse isOpen={isOpen}>
                <Card id="cueCard">
                  <CardBody>
                    <p><strong>Author:</strong> {item.author}</p>
                    <br />
                    <p><strong>Topic:</strong> {item.topic}</p>
                    <br />
                    <p><strong>Description:</strong></p>
                    <p>{item.description}</p>
                    <br />
                    <p><strong>Urgency:</strong> {Urgencylevel}</p>
                    <br />
                    <p><strong>Date created:</strong> {hours} : {mins}</p>

                    <Button disabled={markasdoneDisabled} color="success" className="queueBtnBlock" onClick={() => mark()}>Mark as done</Button>
                    <Button color="warning" disabled={disabled} className="queueBtnBlock" onClick={toggle}>Update ticket</Button>
                    <Button color="danger" disabled={disabled} className="queueBtnBlock" onClick={() => deleteT()}>Delete ticket</Button>


                    {/* UPDATE MODAL */}
                    <div>
                      <Modal isOpen={modal} toggle={toggle} className={className}>
                        <ModalHeader toggle={toggle}>Update ticket</ModalHeader>
                        <Form onSubmit={handleSubmit}>
                          <ModalBody>  
                            <InputGroup>
                              {checkAuth}                
                            </InputGroup>
                            <br />
                            <InputGroup>
                              {checkTitle}
                            </InputGroup>
                            <br />
                            <InputGroup>
                              {checkDesc}
                            </InputGroup>
                            <br />
                            <FormGroup>
                              <Label for="radioLabel">Topic</Label>
                              <div>
                                <CustomInput type="radio" id="Dev Ops" name="topic" onChange={(e) => setTopic(e.target.value)} value="Dev Ops" label="Dev Ops" defaultChecked={TopicCheck1}/>
                                <CustomInput type="radio" id="General" name="topic" onChange={(e) => setTopic(e.target.value)} value="General" label="General" defaultChecked={TopicCheck2}/>
                                <CustomInput type="radio" id="Back-end" name="topic" onChange={(e) => setTopic(e.target.value)} value="Back-end" label="Back-end" defaultChecked={TopicCheck3}/>
                                <CustomInput type="radio" id="Front-end" name="topic" onChange={(e) => setTopic(e.target.value)} value="Front-end" label="Front-end" defaultChecked={TopicCheck4}/>
                                <CustomInput type="radio" id="Software" name="topic" onChange={(e) => setTopic(e.target.value)} value="Software" label="Software" defaultChecked={TopicCheck5}/>
                              </div>
                              {topicCheck}
                            </FormGroup>
                            <FormGroup>
                              <Label for="radioLabel">Urgency</Label>
                              <div>
                                <CustomInput type="radio" id="exampleCustomRadio" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="1" label="Most urgent" defaultChecked={UrgencyCheck1}/>
                                <CustomInput type="radio" id="exampleCustomRadio2" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="2" label="Very urgent" defaultChecked={UrgencyCheck2}/>
                                <CustomInput type="radio" id="exampleCustomRadio3" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="3" label="Slightly urgent" defaultChecked={UrgencyCheck3}/>
                                <CustomInput type="radio" id="exampleCustomRadio4" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="4" label="Less urgent" defaultChecked={UrgencyCheck4}/>
                                <CustomInput type="radio" id="exampleCustomRadio5" onChange={(e) => setUrgency(e.target.value)} name="urgency" value="5" label="Least urgent" defaultChecked={UrgencyCheck5}/>
                              </div>
                              {urgencyCheck}
                            </FormGroup>
                            <Input type="hidden" name="completed" id="completed" value="false"/>
                            </ModalBody>
                            <ModalFooter>
                            <Button type="submit" disabled={!isUpdateEnabled} color="primary" onClick={toggle}>Update ticket</Button>
                            <Button color="secondary" onClick={toggle}>Cancel</Button>
                            </ModalFooter>
                        </Form>
                          
                      </Modal>
                    </div>


                  </CardBody>
                </Card>
              </Collapse>
          </div>
    )

}
export default Ticket;