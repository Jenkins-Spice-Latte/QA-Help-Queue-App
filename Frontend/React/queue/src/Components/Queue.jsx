import { useState } from 'react';
import axios from 'axios';
import Ticket from './Ticket'

const Queue = (props) => {
  const {className} = props;

  console.log("queue")
  console.log(props.isLoaded)

  const [data, setData] = useState([]);
  
  let topic1data = [];
  let topic2data = [];
  let topic3data = [];
  let topic4data = [];
  let topic5data = [];

  let urgency1data = [];
  let urgency2data = [];
  let urgency3data = [];
  let urgency4data = [];
  let urgency5data = [];

  if(props.topicfilter.includes("Topic1")) {
    topic1data = data.filter(item => item.topic === "Topic1");
  }

  if(props.topicfilter.includes("Topic2")) {
    topic2data = data.filter(item => item.topic === "Topic2");
  }

  if(props.topicfilter.includes("Topic3")) {
    topic3data = data.filter(item => item.topic === "Topic3");
  }

  if(props.topicfilter.includes("Topic4")) {
    topic4data = data.filter(item => item.topic === "Topic4");
  }

  if(props.topicfilter.includes("Topic5")) {
    topic5data = data.filter(item => item.topic === "Topic5");
  }
  
  let topicdata = topic1data.concat(topic2data, topic3data, topic4data, topic5data);

  if(props.urgentfilter.includes(1)) {
    urgency1data = topicdata.filter(item => item.urgency === 1);
  }

  if(props.urgentfilter.includes(2)) {
    urgency2data = topicdata.filter(item => item.urgency === 2);
  }

  if(props.urgentfilter.includes(3)) {
    urgency3data = topicdata.filter(item => item.urgency === 3);
  }

  if(props.urgentfilter.includes(4)) {
    urgency4data = topicdata.filter(item => item.urgency === 4);
  }

  if(props.urgentfilter.includes(5)) {
    urgency5data = topicdata.filter(item => item.urgency === 5);
  }

  let urgencydata = urgency1data.concat(urgency2data, urgency3data, urgency4data, urgency5data);

  const result = urgencydata.filter(item => 
    item.author.toLowerCase().includes(props.authorfilter.toLowerCase())
  );

  if(props.isLoaded === false){
      axios.get("http://localhost:8902/readAll")
      .then(response => {
          console.log(response.data)
          setData(response.data);
          props.switchLoaded(); 
      });
  }

    return (
      <> 
      <div className= "queue_div">
        <p>Pending Tickets</p>
        {props.sort}
        {result.map((item) => {
            if(item.complete === false)
              return <Ticket item={item} className={className} mode={(props.mode)} switchLoaded={props.switchLoaded} isLoaded={props.isLoaded}/>  
          })}
      </div>
        
      <div className= "queue_div">
        <p>Completed Tickets</p>
        {result.map((item) => {
          if(item.complete === true)
            return <Ticket item={item} className={className} switchLoaded={props.switchLoaded} isLoaded={props.isLoaded}/>  
         })}
      </div>
        
      </>
    );
  };
  
  export default Queue;