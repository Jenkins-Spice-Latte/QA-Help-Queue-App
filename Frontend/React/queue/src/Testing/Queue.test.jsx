import axios from 'axios';
import Queue from '../Components/Queue';
import renderer from 'react-test-renderer';
import Ticket from '../Components/Ticket';

test('tests to see if Ticket renders properly', () => {
  let ticket = {
    ticket_id: 1,
    author: "auth",
    complete: true,
    description: "description",
    time_created: 101551120,
    title: "title",
    topic: "Topic1",
    urgency: 1
  };
  const TicketComp = renderer.create(<Ticket item={ticket} className={Queue.className} sort="Oldest" mode="Trainee mode" switchLoaded={false} isLoaded={false}/>).toJSON();
  expect(TicketComp).toMatchSnapshot();
});