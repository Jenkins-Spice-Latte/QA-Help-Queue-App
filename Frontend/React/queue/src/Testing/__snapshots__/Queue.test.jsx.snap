// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`tests to see if Ticket renders properly 1`] = `
<div
  className="ticket_div"
>
  <p
    className="ticket_comp title_comp"
  >
    title
     (
    Topic1
    ) 
  </p>
  <svg
    className="expandQueueIc"
    fill="currentColor"
    height="1em"
    onClick={[Function]}
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M1.646 4.646a.5.5 0 01.708 0L8 10.293l5.646-5.647a.5.5 0 01.708.708l-6 6a.5.5 0 01-.708 0l-6-6a.5.5 0 010-.708z"
      fillRule="evenodd"
    />
  </svg>
  <svg
    className="completedIc"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 512 512"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      d="M504 256c0 136.967-111.033 248-248 248S8 392.967 8 256 119.033 8 256 8s248 111.033 248 248zM227.314 387.314l184-184c6.248-6.248 6.248-16.379 0-22.627l-22.627-22.627c-6.248-6.249-16.379-6.249-22.628 0L216 308.118l-70.059-70.059c-6.248-6.248-16.379-6.248-22.628 0l-22.627 22.627c-6.248 6.248-6.248 16.379 0 22.627l104 104c6.249 6.249 16.379 6.249 22.628.001z"
    />
  </svg>
  <svg
    className="mostUrgIcQ"
    fill="currentColor"
    height="1em"
    stroke="currentColor"
    strokeWidth="0"
    style={
      Object {
        "color": undefined,
      }
    }
    viewBox="0 0 16 16"
    width="1em"
    xmlns="http://www.w3.org/2000/svg"
  >
    <path
      clipRule="evenodd"
      d="M16 8A8 8 0 110 8a8 8 0 0116 0zM8 3.5a.5.5 0 00-1 0V9a.5.5 0 00.252.434l3.5 2a.5.5 0 00.496-.868L8 8.71V3.5z"
      fillRule="evenodd"
    />
  </svg>
  <div
    className="collapse"
    style={Object {}}
  >
    <div
      className="card"
      id="cueCard"
    >
      <div
        className="card-body"
      >
        <p>
          <strong>
            Author:
          </strong>
           
          auth
        </p>
        <br />
        <p>
          <strong>
            Topic:
          </strong>
           
          Topic1
        </p>
        <br />
        <p>
          <strong>
            Description:
          </strong>
        </p>
        <p>
          description
        </p>
        <br />
        <p>
          <strong>
            Urgency:
          </strong>
           
          1
           
          Most urgent
        </p>
        <br />
        <p>
          <strong>
            Date created:
          </strong>
           
          5
           : 
          12
        </p>
        <button
          aria-label={null}
          className="queueBtnBlock btn btn-success disabled"
          disabled={true}
          onClick={[Function]}
          type="button"
        >
          Mark as done
        </button>
        <button
          aria-label={null}
          className="queueBtnBlock btn btn-warning"
          onClick={[Function]}
          type="button"
        >
          Update ticket
        </button>
        <button
          aria-label={null}
          className="queueBtnBlock btn btn-danger"
          onClick={[Function]}
          type="button"
        >
          Delete ticket
        </button>
        <div />
      </div>
    </div>
  </div>
</div>
`;
